controllers/cluster_controller.go

```diff
@@ -257,32 +267,43 @@ func (r *ClusterReconciler) reconcileDelete(ctx context.Context, cluster *cluste
 	}
 
 	if cluster.Spec.InfrastructureRef != nil {
-		obj, err := external.Get(ctx, r.Client, cluster.Spec.InfrastructureRef, cluster.Namespace)
-		switch {
-		case apierrors.IsNotFound(errors.Cause(err)):
-			// All good - the infra resource has been deleted
-		case err != nil:
-			return ctrl.Result{}, errors.Wrapf(err, "failed to get %s %q for Cluster %s/%s",
-				path.Join(cluster.Spec.InfrastructureRef.APIVersion, cluster.Spec.InfrastructureRef.Kind),
-				cluster.Spec.InfrastructureRef.Name, cluster.Namespace, cluster.Name)
-		default:
-			// Issue a deletion request for the infrastructure object.
-			// Once it's been deleted, the cluster will get processed again.
-			if err := r.Client.Delete(ctx, obj); err != nil {
-				return ctrl.Result{}, errors.Wrapf(err,
-					"failed to delete %v %q for Cluster %q in namespace %q",
-					obj.GroupVersionKind(), obj.GetName(), cluster.Name, cluster.Namespace)
-			}
-
-			// Return here so we don't remove the finalizer yet.
-			return ctrl.Result{}, nil
+		err := r.deleteObjectViaReference(ctx, cluster, cluster.Spec.InfrastructureRef)
+		// The only way to pass this is to try and get the ref object and find it is gone.
+		if !apierrors.IsNotFound(errors.Cause(err)) {
+			return ctrl.Result{}, err
 		}
 	}
 
 	cluster.Finalizers = util.Filter(cluster.Finalizers, clusterv1.ClusterFinalizer)
 	return ctrl.Result{}, nil
 }
 
+func (r *ClusterReconciler) deleteObjectViaReference(ctx context.Context, cluster *clusterv1.Cluster, ref *corev1.ObjectReference) error {
+	// nothing to delete
+	if ref == nil {
```

Comment:
this check is not necessary based on the fact that all calls to this function are guarded by a nil check